test/integration/diff_drive_system.cc

```diff
@@ -310,6 +310,97 @@ TEST_P(DiffDriveTest, SkidPublishCmd)
   EXPECT_LT(poses[0].Rot().Z(), poses[3999].Rot().Z());
 }
 
+/////////////////////////////////////////////////
+TEST_P(DiffDriveTest, OdomFrameId)
+{
+  // Start server
+  ServerConfig serverConfig;
+  serverConfig.SetSdfFile(std::string(PROJECT_SOURCE_PATH) +
+      "/test/worlds/diff_drive.sdf");
+
+  Server server(serverConfig);
+  EXPECT_FALSE(server.Running());
+  EXPECT_FALSE(*server.Running(0));
+
+  server.SetUpdatePeriod(0ns);
+
+  unsigned int odomPosesCount = 0;
+  std::function<void(const msgs::Odometry &)> odomCb =
+    [&odomPosesCount](const msgs::Odometry &_msg)
+    {
+      ASSERT_TRUE(_msg.has_header());
+      ASSERT_TRUE(_msg.header().has_stamp());
+
+      ASSERT_GT(_msg.header().data_size(), 1);
+
+      EXPECT_STREQ(_msg.header().data(0).key().c_str(), "frame_id");
+      EXPECT_STREQ(
+            _msg.header().data(0).value().Get(0).c_str(), "vehicle/odom");
+
+      EXPECT_STREQ(_msg.header().data(1).key().c_str(), "child_frame_id");
+      EXPECT_STREQ(
+            _msg.header().data(1).value().Get(0).c_str(), "vehicle/chassis");
+
+      odomPosesCount++;
+    };
+
+  transport::Node node;
+  auto pub = node.Advertise<msgs::Twist>("/model/vehicle/cmd_vel");
+  node.Subscribe("/model/vehicle/odometry", odomCb);
+
+  msgs::Twist msg;
+  msgs::Set(msg.mutable_linear(), math::Vector3d(0.5, 0, 0));
+  msgs::Set(msg.mutable_angular(), math::Vector3d(0.0, 0, 0.2));
+
+  pub.Publish(msg);
+
+  server.Run(true, 100, false);
+
+  EXPECT_EQ(5u, odomPosesCount);
```

Comment:
Looks like there are some tests failing due to a timing issue. What we've done in the past when waiting for messages to be published is to add a check with a timeout. Here's an example:
https://github.com/ignitionrobotics/ign-gazebo/blob/65b4b786f75e4517e38c7ccd55238c1e0173406d/test/integration/diff_drive_system.cc#L293-L303